### PR TITLE
feat(core): added stats to track pubdata needed for compression

### PR DESCRIPTION
**Pull request by koloz193**
# What ❔

Updated code responsible for determining seal criteria based on pubdata size to take into account new compression strategy. 

Code does NOT affect the flow or current values.

## Why ❔

With our new compression strategy, we can no longer assume the size of initial/repeated writes. Calculating based on actual value updates will allow us to fit more writes into a batch.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: core/lib/constants/src/ethereum.rs
Comment (perekopskiy) [inline core/lib/constants/src/ethereum.rs:15]: isn't margin equal to 128 - 120 = 8?

File: core/lib/types/src/storage_writes_deduplicator.rs
Comment (perekopskiy) [inline core/lib/types/src/storage_writes_deduplicator.rs:16]: nit: I'd prefer to change field name to `compressed_size` as it is more explicit

**Review by perekopskiy — COMMENTED** [review]
> Code does NOT affect the flow or current values, we'll need to update [size fn](https://github.com/matter-labs/zksync-era/blob/b30f9f81b42771179c352dc137d5f0bf57e0fb02/core/lib/types/src/tx/tx_execution_info.rs#L80) in tx_execution_info for this to be in use.

We need and already can make `size fn` logic depending on protocol version. If you wish it can be a separate PR but anyway it needs to be done.

File: core/lib/types/src/tx/tx_execution_info.rs
Comment (mm-zk) [inline core/lib/types/src/tx/tx_execution_info.rs:38]: comment

File: core/lib/types/src/storage_writes_deduplicator.rs
Comment (mm-zk) [inline core/lib/types/src/storage_writes_deduplicator.rs:435]: can we check metrics values inside these tests?

**Review by mm-zk — APPROVED** [review]

**Comment by koloz193** [comment]
> > Code does NOT affect the flow or current values, we'll need to update [size fn](https://github.com/matter-labs/zksync-era/blob/b30f9f81b42771179c352dc137d5f0bf57e0fb02/core/lib/types/src/tx/tx_execution_info.rs#L80) in tx_execution_info for this to be in use.
> 
> We need and already can make `size fn` logic depending on protocol version. If you wish it can be a separate PR but anyway it needs to be done.

I will follow up with a separate PR covering this

File: core/lib/types/src/tx/tx_execution_info.rs
Comment (perekopskiy) [inline core/lib/types/src/tx/tx_execution_info.rs:66]: It's better to do `if protocol_version >= 17 then ... else ...`, so for versions > 17 we apply new impl

File: core/lib/zksync_core/src/gas_tracker/mod.rs
Comment (perekopskiy) [inline core/lib/zksync_core/src/gas_tracker/mod.rs:117]: you panic here because we don't have `compressed_state_diffs` merged into main, right?

Comment (perekopskiy) [reply core/lib/zksync_core/src/gas_tracker/mod.rs:117]: anyway, the whole expression looks messy, please store the result of the match in a separate variable

File: core/lib/zksync_core/src/state_keeper/seal_criteria/conditional_sealer.rs
Comment (perekopskiy) [inline core/lib/zksync_core/src/state_keeper/seal_criteria/conditional_sealer.rs:37]: This is incorrect to use `latest()` here. I'd make `find_unexecutable_reason` accept version id as argument. It's called only in a `tx_sender` where it's okay to assume that the current protocol version is equal to the protocol versions of the latest sealed miniblock

File: core/lib/zksync_core/src/gas_tracker/mod.rs
Comment (koloz193) [reply core/lib/zksync_core/src/gas_tracker/mod.rs:117]: correct, i can update it. i tried out the match with addition and thought it was a clever way of chaining the two but dont feel strongly about keeping it that way

File: core/lib/zksync_core/src/state_keeper/seal_criteria/conditional_sealer.rs
Comment (perekopskiy) [reply core/lib/zksync_core/src/state_keeper/seal_criteria/conditional_sealer.rs]: I meant not just moving `ProtocolVersionId::latest` stack up but also replacing it with "protocol versions of the latest sealed miniblock" that you can load from DB

File: core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs
Comment (perekopskiy) [inline core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs:96]: Here the approach should be the same as for `find_unexecutable_reason`

File: core/lib/zksync_core/src/state_keeper/seal_criteria/conditional_sealer.rs
Comment (perekopskiy) [reply core/lib/zksync_core/src/state_keeper/seal_criteria/conditional_sealer.rs]: @koloz193

File: core/lib/zksync_core/src/api_server/tx_sender/mod.rs
Comment (perekopskiy) [inline core/lib/zksync_core/src/api_server/tx_sender/mod.rs:923]: nit, you can unwrap in both places safely:
1. Unwrapping header is safe because genesis header must be present before running API.
2. Protocol versions are not set only for some historical blocks, so it's safe to unwrap `protocol_version` of **latest** block

Then you won't need `MiniblockHeader` to implement `Default`.

File: core/lib/types/src/fee.rs
Comment (RomanBrodetski) [inline core/lib/types/src/fee.rs:27]: nit: having a comment is sufficient but in theory we could name it  `updated_values_bytes_size`. Feel free to discard.

File: core/lib/types/src/storage_writes_deduplicator.rs
Comment (RomanBrodetski) [inline core/lib/types/src/storage_writes_deduplicator.rs:209]: ah oh this method is complex, and we adding more mutable/imperative statements.. I don't see a quick improvement here - just something to keep in mind in the future

File: core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs
Comment (RomanBrodetski) [inline core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs:115]: Having to pass `protovol_version ` everywhere is a little awkward. I feel like it's worth to discuss a little if there are other approaches. If we are not mindful about this, we might en d up having a `protocol_version` expanding to every method in our codebase. 

@popzxc @slowli - any quick thoughts on this?

**Review by RomanBrodetski — COMMENTED** [review]
I really like how we develop this feature in small PRs. I can follow them - unlike having a long-lived branch

File: core/lib/types/src/fee.rs
Comment (koloz193) [reply core/lib/types/src/fee.rs:27]: okay, i think thats a fair point. if another update is needed before merging in then i will include this as well

File: core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs
Comment (koloz193) [reply core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs:115]: could we instead have it as part of the `StateKeeperConfig`, initially i tried to keep it out of there because i thought it was protocol version agnostic but seeing `bootloader_hash` and `default_aa_hash`, i'm thinking it has some reliance on protocol version. wdyt?

Comment (perekopskiy) [reply core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs:115]: `StateKeeperConfig` is static, meaning it's loaded from env at the start and isn't changed, whilst protocol version can be changed. `bootloader_hash` and `default_aa_hash` are artifacts, hopefully, they are not used anywhere and should be removed.

As for Roman's concern: I generally agree, maybe making seal criteria depend on protocol version is not the best solution. But we can remove this dependency after the upgrade, so it doesn't look that bad

Comment (popzxc) [reply core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs:115]: Agree with Misha (though I'd mark it with a TODO to actually remove it one day).

File: core/lib/zksync_core/src/gas_tracker/mod.rs
Comment (popzxc) [inline core/lib/zksync_core/src/gas_tracker/mod.rs:117]: (I didn't fully understand the previous discussion on this line, so sorry in advance): do I get it right that this panic is going to be fixed soon? If so, would it make sense to replace `panic!` with `todo!` to make the intention more explicit?

Comment (perekopskiy) [reply core/lib/zksync_core/src/gas_tracker/mod.rs:117]: Also, this panic makes upgrade test fail in CI. I don't like this test to rely on the fact that server works with an incremented version, but I also don't see how it can be changed.. I think the easiest solution is to wait for https://github.com/matter-labs/zksync-era/pull/259 to be merged and remove this panic

File: core/lib/types/src/storage_writes_deduplicator.rs
Comment (slowli) [inline core/lib/types/src/storage_writes_deduplicator.rs:113]: Maybe a stupid question: Could `metrics` be computed on demand based on `self.modified_key_values` instead of being contained in `StorageWritesDeduplicator`? IMO, if it's possible (and doesn't have significant computational overhead), it would be preferable because such computation would be stateless (doesn't need to take rollbacks into account etc.).

File: core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs
Comment (slowli) [reply core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs:115]: In general, we could have the set of criteria used for a certain input be dependent on the protocol version (e.g., by extending the `SealCriterion` interface). In this way, dependency on the protocol version would be contained in `ConditionalSealer`. I'm not sure whether this would make things better or worse, given that it seems like this dependency on the protocol version isn't supposed to be permanent.

File: core/lib/types/src/block.rs
Comment (koloz193) [reply core/lib/types/src/block.rs:68]: yea, we changed from `unwrap_or_default` to `unwrap` in the place it should have been used. will update

File: core/lib/zksync_core/src/api_server/tx_sender/mod.rs
Comment (koloz193) [reply core/lib/zksync_core/src/api_server/tx_sender/mod.rs:923]: good call out, both calling functions can be refactored a bit to passthrough a StorageProcessor

**Review by popzxc — DISMISSED** [review]
Disclaimer: I only checked the code, not the domain logic.
Strongly recommend getting approval from someone with enough knowledge to validate the correctness 😅

File: core/lib/types/src/protocol_version.rs
Comment (StanislavBreadless) [inline core/lib/types/src/protocol_version.rs:76]: nit: shall we have `todo` here too?

**Review by StanislavBreadless — APPROVED** [review]

File: core/lib/types/src/protocol_version.rs
Comment (koloz193) [reply core/lib/types/src/protocol_version.rs:76]: we could, originally @perekopskiy said we should use what v17 was using but if its going to be changed with the overall boojum integration merge i think todo is safe

Comment (perekopskiy) [reply core/lib/types/src/protocol_version.rs:76]: we can't use todo because of upgrade test..

**Comment by koloz193** [comment]
closing this pr in favor of returning the data from the vm

**Pull request closed without merging**
